### DISCUSSION TITLE
RC Car

### DISCUSSION THREAD

**Discussion by rodo54** [micropython/micropython#9925] [Show and tell]
Hi everyone,

I am sharing this project of a  [simple Remote Control Car](https://github.com/rodo54/Esp32_Car)
Although it is its first version, it is already fully functional.
The page includes the codes, description and diagrams for its realization.
**All comments, improvements and contributions are welcome.**
Feel free to contact me for further info.

Rodolfo
rleibner@gmail.com

**Comment by jimmo** [comment]
Very cool!

Great to see you using the accelerometer too. I did a similar thing for my son using a micro:bit as the controller (which has an on-board accelerometer) because his hands were too small to hold a regular remote control. We had a blast with it! It was great to be able to teach him to drive and now he's older he can use a full-size controller really well.

It did take some tweaking to make it feel comfortable to drive with the accelerometer though though... I think one thing was to adjust the sensitivity based on the current speed.

**Reply by rodo54** [reply]
What wonderful children! In my case it was my grandson -I'm teaching him python- who suggested this project to me.
Of course, I was enthusiastic and we are enjoying it together (me with the difficulties of the design, and he with its use).
My first idea was to use the analog values of the accelerometer to command both the motor and the servo.
In practice, it's not easy to drive like this...
That is why I am using the accelerometer digitally, that is to say to produce "increase/decrement" instructions.

> I think one thing was to adjust the sensitivity based on the current speed.

Did that work?
Can you share the idea or pseudo-code?

**Reply by jimmo** [reply]
@rodo54 Yes we found controlling speed with the accelerometer was basically impossible. Especially without doing some extra maths to compensate for the tilt due to the steering at the same time, so using buttons was better for speed. We went through a few different versions but what worked best was a "go" button and a "boost" button (i.e. hold "go" to drive, and tap boost when you want it). I implemented a basic acceleration curve with the boost button to prevent wheel spin, and then it fades off. (i.e. you can't just hold boost down, you have to press it again for another boost -- it's kind of fun that way).

> Did that work?

It worked well enough. The basic problem is that at speed the steering is just so much more sensitive, but he would still fling the controller through 90 degrees. So introducing some non-linearity proportional to the current speed (especially if boosting) helped with that. The other thing that didn't work well at first was that of course rotating past 90 degrees caused the steering to go backwards again. So finding a way to detect when the phone was rotated too far (i.e. check when the up/down axis was inverted).

It would be heaps better to try and get a more accurate idea of the phone orientation -- this is very difficult to do with the accelerometer alone. I bet on a phone there are all sorts of libraries for sensor fusion to make this easier.

**Comment by rodo54** [comment]
Now I'm trying to do a "stand-alone" Controller using another ESP32, a mini-joystick and an oled display.
Following the published examples ([simple_central](https://github.com/micropython/micropython/blob/master/examples/bluetooth/ble_simple_central.py) and [simple_peripheral](https://github.com/micropython/micropython/blob/master/examples/bluetooth/ble_simple_peripheral.py) I had no difficulties to communicate from the Controller to the Car.

But I'd like to read the voltage of the car batteries on the display. And for that I need to add a communication from the car to the controller. 
And I have not been able to solve it.

**Reply by jimmo** [reply]
If the car is the GATT server (typically the peripheral) then you can add another characteristic which you periodically write the battery voltage to. Then the client can either subscribe to it, or periodically poll (read) it.

If you're using BLE, then it's well worth looking at `aioble` https://github.com/micropython/micropython-lib/tree/master/micropython/bluetooth/aioble as it makes all of this much simpler.
